File: test/features/widget_config/domain/widget_route_test.dart
import 'dart:convert';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_railway_timetable/features/widget_config/domain/entity/widget_route.dart';

void main() {
  group('WidgetRoute.toJson', () {
    test('encodes TR route correctly', () {
      const route = WidgetRoute(
        system: WidgetRailwaySystem.tr,
        fromId: '1000',
        fromName: '臺北',
        toId: '4400',
        toName: '高雄',
      );
      final json = jsonDecode(route.toJson()) as Map<String, dynamic>;
      expect(json['system'], 'TR');
      expect(json['fromId'], '1000');
      expect(json['fromName'], '臺北');
      expect(json['toId'], '4400');
      expect(json['toName'], '高雄');
    });

    test('encodes HSR route correctly', () {
      const route = WidgetRoute(
        system: WidgetRailwaySystem.hsr,
        fromId: '1000',
        fromName: '台北',
        toId: '1070',
        toName: '左營',
      );
      final json = jsonDecode(route.toJson()) as Map<String, dynamic>;
      expect(json['system'], 'HSR');
      expect(json['toId'], '1070');
    });

    test('produces valid JSON string', () {
      const route = WidgetRoute(
        system: WidgetRailwaySystem.tr,
        fromId: '1000',
        fromName: '臺北',
        toId: '4400',
        toName: '高雄',
      );
      expect(() => jsonDecode(route.toJson()), returnsNormally);
    });

    test('system field matches iOS RailwaySystem rawValue', () {
      // iOS WidgetRoute decodes "TR" / "HSR" via RailwaySystem.init(rawValue:)
      final trJson = jsonDecode(const WidgetRoute(
        system: WidgetRailwaySystem.tr,
        fromId: '1000', fromName: '臺北',
        toId: '4400', toName: '高雄',
      ).toJson()) as Map<String, dynamic>;
      final hsrJson = jsonDecode(const WidgetRoute(
        system: WidgetRailwaySystem.hsr,
        fromId: '1000', fromName: '台北',
        toId: '1070', toName: '左營',
      ).toJson()) as Map<String, dynamic>;

      expect(trJson['system'], 'TR');
      expect(hsrJson['system'], 'HSR');
    });
  });
}

File: android/app/src/test/kotlin/com/example/flutter_railway_timetable/widget/domain/entity/WidgetRouteTest.kt
package com.example.flutter_railway_timetable.widget.domain.entity

import org.junit.Assert.assertEquals
import org.junit.Test

class WidgetRouteTest {

    @Test
    fun `defaultFor TR returns 臺北 to 高雄`() {
        val route = WidgetRoute.defaultFor("TR")
        assertEquals("臺北", route.fromName)
        assertEquals("1000", route.fromId)
        assertEquals("高雄", route.toName)
        assertEquals("4400", route.toId)
        assertEquals("TR", route.system)
    }

    @Test
    fun `defaultFor HSR returns 臺北 to 左營`() {
        val route = WidgetRoute.defaultFor("HSR")
        assertEquals("臺北", route.fromName)
        assertEquals("1000", route.fromId)
        assertEquals("左營", route.toName)
        assertEquals("1070", route.toId)
        assertEquals("HSR", route.system)
    }

    @Test
    fun `defaultFor unknown system falls back to TR`() {
        val route = WidgetRoute.defaultFor("UNKNOWN")
        assertEquals("TR", route.system)
    }
}
